### PR TITLE
IEN-110 Import applicant data and fix add/update/filter applicant API

### PR DESCRIPTION
**Pull request by bhumin-fw**
IEN-110 Import applicant data and fix add/update/filter applicant API based on the new schema

### CODE REVIEW

File: apps/api/src/applicant/dto/ienapplicant-create.dto.ts [hunk 1]
```diff
@@ -8,73 +8,112 @@ import {
...
 
-  @ApiPropertyOptional({ description: 'Applicant citizenship', default: 'Maxico' })
+  @ApiPropertyOptional({
+    description: 'Applicant phone number',
+    default: '77-555-1234',
+  })
   @IsOptional()
-  citizenship?: string;
+  @IsString()
+  @Length(1, 256, { message: 'Please provide applicant phone' })
```
Comment (coryatwater-FW) [inline apps/api/src/applicant/dto/ienapplicant-create.dto.ts:46]: don't think it's too big of a deal, but I think the max digits a phone number can have is 14 or 15? maybe cutting it down to something closer would help a user notice any errors more easily

File: apps/api/src/applicant/dto/ienapplicant-create.dto.ts [hunk 2]
```diff
@@ -8,73 +8,112 @@ import {
...
+  @IsOptional()
+  pr_status?: string;
 
   @ApiPropertyOptional({
-    description: 'Applicant education, put it in comma separated',
-    default: 'PhD, Master of Nursing',
+    description: 'Nursing educations',
+    default: [
+      { name: 'PhD', coutry: 'ca' },
+      { name: 'Master of Nursing', coutry: 'ca' },
```
Comment (coryatwater-FW) [inline apps/api/src/applicant/dto/ienapplicant-create.dto.ts:83]: typo on country

File: apps/api/src/applicant/external-api.service.ts [hunk 3]
```diff
@@ -105,4 +108,101 @@ export class ExternalAPIService {
...
     }
   }
+
+  /**
+   * fetch and upsert applicant details
+   */
+  async saveApplicant(): Promise<void> {
+    try {
+      const data = await this.external_request.getApplicants();
+      await this.createBulkApplicans(data);
```
Comment (coryatwater-FW) [inline apps/api/src/applicant/external-api.service.ts:118]: typo on createBulkApplicants

File: apps/api/src/applicant/external-api.service.ts [hunk 4]
```diff
@@ -105,4 +108,101 @@ export class ExternalAPIService {
...
+
+  /**
+   * fet user/staff and HA details to reduce database calls.
+   * @returns
+   */
+  async getApplicantMasterData() {
+    // fetch user/staff details
+    const usersArray = await this.ienUsersRepository.find();
+    const users: any = {};
+    usersArray.map(user => {
```
Comment (sunghwan-park-fw) [inline apps/api/src/applicant/external-api.service.ts:132]: map return an array but it isn't used. how about `forEach` instead of `map`?

Comment (bhumin-fw) [reply apps/api/src/applicant/external-api.service.ts]: forEach is correct here, Just fix it and pushed new code.

File: apps/api/src/applicant/dto/ienapplicant-filter.dto.ts [hunk 5]
```diff
@@ -10,12 +10,12 @@ export class IENApplicantFilterAPIDTO extends IENApplicantFilterDTO {
...
-  })
-  @IsOptional()
-  @IsString()
-  status?: string;
+  // @ApiPropertyOptional({
+  //   description: 'Provide optional status(with comma separated values e.g. 9,10,11)',
+  // })
+  // @IsOptional()
+  // @IsString()
+  // status?: string;
```
Comment (sunghwan-park-fw) [inline apps/api/src/applicant/dto/ienapplicant-filter.dto.ts:18]: This field would be used in the main page. I am just curious why you commented out this.

Comment (bhumin-fw) [reply apps/api/src/applicant/dto/ienapplicant-filter.dto.ts:18]: Some queries are there that need to be solved about status/milestone import data for applicants. So I might need to change its schema and relation, so it won't work with the existing setup. So I have disabled the status/milestone filter, for now,

Comment (bhumin-fw) [reply apps/api/src/applicant/dto/ienapplicant-filter.dto.ts:18]: I'll try to re-enable it early next week. Mostly next Monday itself.

**Review by sunghwan-park-fw — APPROVED** [review]

File: apps/api/src/applicant/ienapplicant.util.service.ts [hunk 6]
```diff
@@ -260,6 +262,7 @@ export class IENApplicantUtilService {
    * @returns
    */
   async getUserArray(users: any): Promise<IENUsers | any> {
+    users = users.map((item: { id: any }) => item.id);
```
Comment (dbayly-freshworks) [inline apps/api/src/applicant/ienapplicant.util.service.ts:265]: This has a lot of `any`, we should ideally have a type here.

File: apps/api/src/applicant/ienapplicant.util.service.ts [hunk 7]
```diff
@@ -48,24 +48,24 @@ export class IENApplicantUtilService {
...
-    if (status) {
-      const status_list = await this.fetchChildStatusList(status);
-      where = { status: In(status_list), ...where };
-      isWhere = true;
-    }
+    // if (status) {
+    //   const status_list = await this.fetchChildStatusList(status);
+    //   where = { status: In(status_list), ...where };
+    //   isWhere = true;
+    // }
```
Comment (dbayly-freshworks) [inline apps/api/src/applicant/ienapplicant.util.service.ts:68]: Commented out code here

File: apps/api/src/applicant/ienapplicant.service.ts [hunk 8]
```diff
@@ -142,25 +138,34 @@ export class IENApplicantService {
...
   async updateApplicantInfo(
     id: string,
-    applicantUpdate: IENApplicantUpdateAPIDTO,
+    applicantUpdate: IENApplicantCreateUpdateAPIDTO,
   ): Promise<IENApplicant | any> {
     const applicant = await this.getApplicantById(id);
-    const { ha_pcn, assigned_to, updated_by, ...data } = applicantUpdate;
-    if (ha_pcn && ha_pcn.length) {
-      applicant.ha_pcn = await this.ienapplicantUtilService.getHaPcn(ha_pcn);
+    // const { ha_pcn, assigned_to, updated_by, ...data } = applicantUpdate;
```
Comment (dbayly-freshworks) [inline apps/api/src/applicant/ienapplicant.service.ts:144]: Commented out code here

**Review by dbayly-freshworks — COMMENTED** [review]
This is pretty much there, but needs some clean up.

File: apps/api/src/applicant/dto/ienapplicant-filter.dto.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (bhumin-fw) [reply apps/api/src/applicant/dto/ienapplicant-filter.dto.ts:18]: It's temporary, I'll enable it in the next couple of days, once the milestone/status import completes successfully.

File: apps/api/src/applicant/ienapplicant.util.service.ts (hunk 7 above)
Comment (bhumin-fw) [reply apps/api/src/applicant/ienapplicant.util.service.ts:68]: It's temporary, I'll enable it in the next couple of days, once the milestone/status import completes successfully.

File: apps/web/src/pages/details/[applicantId].tsx [hunk 9]
```diff
@@ -74,37 +76,44 @@ const Details = () => {
...
         </div>
         <div className='col-span-3'>
-          <DetailsItem title='Country of Nursing Education' text={applicant.country_of_training} />
+          <DetailsItem title='Country of Nursing Education' text='Needs to remove it' />
         </div>
         <div className='col-span-3'>
-          <DetailsItem title='Nursing Education' text={applicant.education} />
+          <DetailsItem
+            title='Nursing Education'
+            text={applicant.nursing_educations.map((a: { name: any }) => a.name).join(',')}
```
[the remark this answers is not in the document]
Comment (bhumin-fw) [reply apps/web/src/pages/details/[applicantId].tsx:112]: "any" taken by code formate and quick fix apply, Let me update it to string | number.

**Review by dbayly-freshworks — APPROVED** [review]
Looks good to me

**Pull request merged**